README.md
# Unofficial CSGOEmpire API Wrapper
The unofficial CSGOEmpire API JavaScript Wrapper

You can find the **Official CSGOEmpire Api** documentation [here](https://github.com/OfficialCSGOEmpire/API-Docs#csgoempire-api-key-documentation)

You can also find the change log file [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/CHANGELOG.md)

This wrapper is not complete yet. If you feel like contributing, open a pull request proposing a change :)

## Installation
With npm:
```
npm install csgoempire-api
```
With Yarn:
```
yarn add csgoempire-api
```

## Usage
```javascript
import { CSGOEmpire } from "csgoempire-api"

const empire = new CSGOEmpire(YOUR_API_KEY)

empire.getActiveTrades()
empire.getActiveAuctions()
...
```

Initializing websocket:
```javascript 
import { CSGOEmpire } from "csgoempire-api"

const empire = new CSGOEmpire(YOUR_API_KEY)

empire.initSocket((socket) => {
    socket.on("new_item", (data) => {
        // ...
    })
})

// empire.socket can be undefined
const socket = empire.socket
```
Note: Not running `initSocket` will result in a undefined socket instance
### initSocket(fn)
| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| fn | Function? | - | A callback function that gets executed when the socket is initialized and authenticated |

## Documentation

### Api 
```javascript
const empire = new CSGOEmpire(apiKey)
```
- apiKey: string (required)
- webSocketEnabled: boolean (optional) (true by default)

---

### Socket
Extends [socket.io-client's](https://github.com/socketio/socket.io-client) class. Can be `undefined` if not initialized

```javascript
const socket = empire.socket
```

#### on(event, fn)

`socket.on` is a `io().on` wrapper that provides typings support for CSGOEmpire's events

```javascript
socket.on(event, fn)
```
- event: string (required)
    - "new_item" 
    - "updated_item"
    - "auction_update"
    - "deleted_item"
    - "trade_status"
    - "timesync" 
- fn: Function (required)

Each event function has typings support. 

```javascript
socket.on("new_item", (data) => {
    /**
     * data extends NewItemSocketData interface
     * 
     * Available properties: 
     * { 
     *      id: number,
     *      name: string,
     *      ...
     * }
     */
    const { id, name } = data
})
```

You can check all events example responses [here](https://github.com/OfficialCSGOEmpire/API-Docs#websocket-events)

---

### getMetadata
Returns the user object, which is used to identify via websocket, as well as socket token (authorizationToken) & socket signature (signature) which are used to authenticate on websocket.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| - | - | - | - |

```javascript
empire.getMetadata().then(res => {
    ...
})
```
You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getMetadata.ts)

---

### getActiveTrades
Returns an array of all items currently being deposited or withdrawn by this account. This does not include bids placed on active items until the auction ends.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| - | - | - | - |

```javascript
empire.getActiveTrades().then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getActiveTrades.ts)

---

### getActiveAuctions
Returns an array of all auctions currently being bid on by this account.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| - | - | - | - |

```javascript
empire.getActiveAuctions().then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getActiveAuctions.ts)

---

### updateSettings
Used to update your tradelink and/or Steam API key

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| data | Object | - | An object containing a trade_url (required) and a steam_api_key (optional)  |

```javascript
empire.updateSettings().then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/updateSettings.ts)

---

### Deposits

Deposit related methods

#### getCSGOInventory
Fetch your inventory from steam and caches it to the database for 1 hour.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| invalid | boolean | false | Filters invalid items, defaults to no filtering |

```javascript
empire.getCSGOInventory().then(res => {
    ...
})
```
You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/csgoInventory.ts)

#### getUniqueInfo
Get inspected unique info for items in user inventory. Examples include float/sticker data

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| - | - | - | - |

```javascript
empire.getUniqueInfo().then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getUniqueInfo.ts)


#### createDeposit
Creates an item deposit

Notes: coin_value is in **coin cents**, so 100.01 coins is represented as 10001

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| data | Object | - | An object containing an array of items to deposit |

```javascript
empire.createDeposit({
    items: [
        {
            "id": 3731677704,
            "custom_price_percentage": 32,
            "coin_value": 576811
        }
    ]
}).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/createDeposit.ts)

#### cancelDeposit
Cancels processing deposit without any bids. Once a bid has been placed items are no longer eligible to be cancelled.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| deposit_id | number | - | The deposited item's id |

```javascript
empire.cancelDeposit(28391470).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/cancelDeposit.ts)

#### sellNow
Sells an on going auction item to the current auction highest bidder

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| deposit_id | number | - | The deposited item's id |

```javascript
empire.sellNow(28391470).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/sellNow.ts)

---

### Withdraw
Withdraw related methods

#### getListedItems
Get a list of all items listed on the withdrawals page

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| page | number | - | The page to fetch the data from |
| per_page | number | - | The ammount of items to be fetched per page |
| options | Object? | - | An object containing filtering options |

```javascript
empire.getListedItems(1, 50, { sort: "asc" }).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getListedItems.ts)

#### getDepositorStats
Get the depositing users stats from a unique deposit ID

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| deposit_id | number | - | The deposited item's id |

```javascript
empire.getDepositorStats(28391470).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/depositorStats.ts)


#### createWithdrawal
Withdraw item directly if the auction has expired without being won.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| deposit_id | number | - | The deposited item's id |

```javascript
empire.createWithdrawal(28391470).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/createWithdrawal.ts)

#### placeBid
Place a bid on an auction.

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| deposit_id | number | - | The deposited item's id |
| bid_value | number | - | The ammount to bid |

```javascript
empire.placeBid(28391470, 60).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/placeBid.ts)

---
### Misc

Other api methods

#### getSeeds
Get roulette seeds

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| page | number | - | The page to fetch the data from |
| per_page | number? | 15 | The ammount of items to be fetched per page |

```javascript
empire.getSeeds(1, 20).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getSeeds.ts)

#### getHistory
Returns rolls history

| Option | Type | Default Value | Description |
| :----: | :--: | :----: | :----: | 
| seed | number | - | The seed to fetch the data from |

```javascript
empire.getHistory(2543).then(res => {
    ...
})
```

You can find the response Object interface [here](https://github.com/gustavo-dev/csgoempire-api/blob/master/lib/typings/requests/getHistory.ts)

CHANGELOG.md
# Change Log

## 1.1.0
Features:
- Added getSeeds method ([bfcbed2](https://github.com/gustavo-dev/csgoempire-api/commit/bfcbed2984686a4744fc0bb99474fe76c4611978))
- Added getHistory method ([bfcbed2](https://github.com/gustavo-dev/csgoempire-api/commit/bfcbed2984686a4744fc0bb99474fe76c4611978))

## 1.0.2
Fixes:
- Fixed typings interfaces ([3dabaf4](https://github.com/gustavo-dev/csgoempire-api/commit/3dabaf41280d48c5efab87aa67423938018abe6e))
## 1.0.1
Fixes: 
- Fixed axios interceptors incorrect data type ([eec5d82](https://github.com/gustavo-dev/csgoempire-api/commit/eec5d82cdba681784c5ef0ed5b216f1e538345c4))


lib/CSGOEmpire.ts
import { Axios } from "axios"

import io from "socket.io-client"

import {
    ExtendedSocket,
    UpdateSettingsData,
    UpdateSettingsResponse,
    MetadataResponse,
    ActiveTradesResponse,
    ActiveAuctionsResponse,
    CSGOInventoryResponse,
    UniqueInfoResponse,
    CreateDepositData,
    CreateDepositResponse,
    CancelDepositResponse,
    SellNowResponse,
    ListedItemsResponse,
    ListedItemsData,
    DepositorStatsResponse,
    CreateWithdrawalResponse,
    PlaceBidResponse,
    GetSeedsResponse,
    GetHistoryResponse,
} from "./typings"

export class CSGOEmpire {
    private api: Axios

    // Trade Socket
    private _socket?: SocketIOClient.Socket

    // private notificationsSocket: SocketIOClient.Socket

    constructor(apiKey?: string) {
        this.api = new Axios({
            baseURL: "https://csgoempire.com/api/v2",
            headers: {
                authorization: apiKey ? `Bearer ${apiKey}` : "",
                "content-type": "application/json",
                "user-agent": "API Bot",
            },
        })

        /**
         * Axios adapter require data to be
         * a String, ArrayBuffer, Buffer or Stream
         */

        this.api.interceptors.request.use((config) => {
            if (config.data) config.data = JSON.stringify(config.data)

            return config
        })

        // Convert data to JSON if string
        this.api.interceptors.response.use(
            (response) => {
                if (typeof response.data === "string") {
                    response.data = JSON.parse(response.data)
                }

                return response
            },
            (error) => {
                return Promise.reject(error)
            }
        )
    }

    public initSocket = (cb: (socket: ExtendedSocket) => void = () => {}) => {
        // Debounce
        if (this._socket) return

        this._socket = io("wss://trade.csgoempire.com/trade", {
            transports: ["websocket"],
            path: "/s/",
            secure: true,
            rejectUnauthorized: false,
            reconnection: true,
            // @ts-ignore
            extraHeaders: {
                "User-Agent": "API Bot",
            },
        })

        this._socket.on("connect", async () => {
            // Typescript thinks socket is undefined in this context
            if (!this._socket) throw Error("Socket is not available")

            const { user, socket_signature, socket_token } =
                await this.getMetadata()

            // Log when connected
            console.log(`Connected to websocket`)

            // Emit the data we got earlier to the socket to identify this client as the user
            this._socket.emit("identify", {
                uid: user.id,
                model: user,
                authorizationToken: socket_token,
                signature: socket_signature,
            })

            // Handle the Init event
            this._socket.on("init", (data: any) => {
                if (data && data.authenticated) {
                    console.log(`Successfully authenticated as ${data.name}`)
                    cb(this.socket)
                }
            })
        })
    }

    /**
     * Extended socket instance
     */
    get socket() {
        if (!this._socket) throw Error("Socket is not available")

        const { on, ...rest } = this._socket

        const _socket = { ...rest } as ExtendedSocket

        _socket.on = (event, fn) => {
            // Typescript thinks this could be undefined
            if (!this._socket) throw Error("Socket is not available")

            return this._socket.on(event, fn)
        }

        return _socket
    }

    /**
     * @returns the user object, which is used to identify via websocket,
     * as well as socket token (authorizationToken) & socket signature (signature)
     * which are used to authenticate on websocket.
     */
    public getMetadata = async () => {
        return (await this.api.get<MetadataResponse>("/metadata/socket")).data
    }

    /**
     * @returns an array of all items currently being deposited or withdrawn by this account.
     * This does not include bids placed on active items until the auction ends.
     */
    public getActiveTrades = async () => {
        return (
            await this.api.get<ActiveTradesResponse>("/trading/user/trades")
        ).data
    }

    /**
     * @returns an array of all auctions currently being bid on by this account.
     */
    public getActiveAuctions = async () => {
        return (
            await this.api.get<ActiveAuctionsResponse>("/trading/user/auctions")
        ).data
    }

    /**
     * Used to update your tradelink and/or Steam API key
     */
    public updateSettings = async (data: UpdateSettingsData) => {
        return (
            await this.api.post<UpdateSettingsResponse>(
                "/trading/user/settings",
                data
            )
        ).data
    }

    /**
     * Fetch your inventory from steam and caches it to the database for 1 hour.
     * @param invalid {boolean} true or false
     */
    public getCSGOInventory = async (invalid: boolean = false) => {
        return (
            await this.api.get<CSGOInventoryResponse>(
                `/trading/user/inventory?invalid=${invalid ? "yes" : "no"}`
            )
        ).data
    }

    /**
     * Get inspected unique info for items in user inventory. Examples include float/sticker data
     */
    public getUniqueInfo = async () => {
        return (
            await this.api.get<UniqueInfoResponse>(
                "/trading/user/inventory/unique-info"
            )
        ).data
    }

    /**
     * Creates an item deposit
     *
     * Notes: coin_value is in coin cents, so 100.01 coins is represented as 10001
     */
    public createDeposit = async (data: CreateDepositData) => {
        return (
            await this.api.post<CreateDepositResponse>("/trading/deposit", data)
        ).data
    }

    /**
     * Cancels processing deposit without any bids. Once a bid has been placed items are no longer eligible to be cancelled.
     * @param deposit_id {number} The deposited item's id
     */
    public cancelDeposit = async (deposit_id: number) => {
        return (
            await this.api.post<CancelDepositResponse>(
                `/trading/deposit/${deposit_id}/cancel`
            )
        ).data
    }

    /**
     * Sells an on going auction item to the current auction highest bidder
     * @param deposit_id {number} The deposited item's id
     * @returns
     */
    public sellNow = async (deposit_id: number) => {
        return (
            await this.api.post<SellNowResponse>(
                `/trading/deposit/${deposit_id}/sell`
            )
        ).data
    }

    /**
     * Get a list of all items listed on the withdrawals page
     * @param page {number} The page to fetch the data from
     * @param per_page {number} The ammount of items to be fetched per page
     * @param options {Object} An object containing filtering options
     */
    public getListedItems = async (
        page: number,
        per_page: number,
        options?: ListedItemsData
    ) => {
        let queryString = ""

        if (options) {
            for (const key in options) {
                queryString += `&${key}=${
                    options[key as keyof ListedItemsData]
                }`
            }
        }

        return (
            await this.api.get<ListedItemsResponse>(
                `/trading/items?page=${page}&per_page=${per_page}${queryString}`
            )
        ).data
    }

    /**
     * Get the depositing users stats from a unique deposit ID
     * @param deposit_id {number} The deposited item's id
     */
    public getDepositorStats = async (deposit_id: number) => {
        return (
            await this.api.get<DepositorStatsResponse>(
                `/trading/deposit/${deposit_id}/stats`
            )
        ).data
    }

    /**
     * Withdraw item directly if the auction has expired without being won.
     * @param deposit_id {number} The deposited item's id
     */
    public createWithdawal = async (deposit_id: number) => {
        return (
            await this.api.post<CreateWithdrawalResponse>(
                `/trading/deposit/${deposit_id}/withdraw`
            )
        ).data
    }

    /**
     * Place a bid on an auction.
     * @param deposit_id {number} The deposited item's id
     * @param bid_value {number} The ammount to bid
     */
    public placeBid = async (deposit_id: number, bid_value: number) => {
        return (
            await this.api.post<PlaceBidResponse>(
                `/trading/deposit/${deposit_id}/bid`,
                { bid_value }
            )
        ).data
    }

    /**
     * Returns roulette seed
     * @param page {number} The page to fetch the data from
     * @param per_page {number} The ammount of items to be fetched per page
     */
    public getSeeds = async (page: number, per_page: number = 15) => {
        return (
            await this.api.get<GetSeedsResponse>(
                `/metadata/roulette/seeds?per_page=${per_page}&page=${page}`
            )
        ).data
    }

    /**
     * Returns rolls history
     * @param seed {number} The seed to fetch the data from
     */
    public getHistory = async (seed: number) => {
        return (
            await this.api.get<GetHistoryResponse>(
                `/metadata/roulette/history?seed=${seed}`
            )
        ).data
    }
}
